### PR TITLE
sphinx generated docs and doc source is added

### PR DESCRIPTION
**Pull request by AhmetCanSolak**
# Description

Specdocs autogenerated from docstrings with help of sphinx and napoloen extension.
This is not final version of PR, please provide review but do not merge. I believe i need to update `.gitignore` file too with this PR.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# References

https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [X] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by jni** [comment]
typically the build products should not be committed to the history?

**Comment by AhmetCanSolak** [comment]
> typically the build products should not be committed to the history?

yes @jni i put all for now but wondering if we want to serve on github or somewhere else? depending on that will modify before merge. How would you like to have these autogenerated docs? local or served? if served where? :D

**Comment by AhmetCanSolak** [comment]
indeed, i realized my mistake independant of where we will serve :D i need to get rid of doctrees, and push only html :D

**Comment by AhmetCanSolak** [comment]
@jni now removed the doctrees folder :)

**Review by kevinyamauchi — COMMENTED** [review]
Thanks! I think the docs look good. I checked the markers layer and they look complete. I've made a few comments below.
* I think the name in the main title on the index.html page ("Welcome to gui’s documentation!") should be napari-gui (or something more related to napari) rather than just "gui".
* I think it would be nice to have a theme with a bit more contrast between different types of elements because I think it's a bit difficult to browse/parse at the moment. I like the [read the docs theme](http://www.writethedocs.org/guide/tools/sphinx-themes/), but I'm open to others as well.
*  We may want to add sphinx to the requirements (docs don't build for me at the moment)

File: docs/build/html/_modules/gui/_version.html
Comment (kne42) [inline docs/build/html/_modules/gui/_version.html]: This file should be ignored.

File: docs/build/html/_modules/gui/_vispy/scene/visuals.html
Comment (kne42) [inline docs/build/html/_modules/gui/_vispy/scene/visuals.html]: IMHO all backports should also be hidden for the time being.

File: docs/build/html/_sources/gui.layers.rst.txt
Comment (kne42) [inline docs/build/html/_sources/gui.layers.rst.txt]: File names with one or more leading underscores should be ignored.

**Review by kne42 — COMMENTED** [review]
Great start to the docs @AhmetCanSolak! Left a few comments about decision-making surrounding what should and should not be included in the generated output.

**Comment by sofroniewn** [comment]
Have just started looking over the docs, and want to make sure I'm doing it right - I navigated to the docs file and then ran `make html` that then generates a bunch of html files in the `build/html` folder. I then opened up `gui.html` and started clicking around. It looks like this for me
image

File: docs/build/html/_modules/gui/_vispy/scene/visuals.html
Comment (sofroniewn) [reply docs/build/html/_modules/gui/_vispy/scene/visuals.html]: agree the vispy backports should be hidden

File: docs/build/html/_sources/gui.layers.rst.txt
Comment (sofroniewn) [reply docs/build/html/_sources/gui.layers.rst.txt]: not sure about hiding all filenames with a leading `_` - or maybe some of our files that have a `_` at the beginning really shouldn't - i.e. `_image_layer.py` and `_markers_layer.py` in the main `layers` folder. I think these should be easily accessible to users as they contain methods that users will interact with, right?

It would also be nice if there was a way to see the files with `_` even if just as part of "developer" docs

**Comment by sofroniewn** [comment]
I agree with all of @kevinyamauchi points above. It would be much nicer to have `napari-gui` everywhere instead of `gui`, maybe this is something that will change / be easier with the namespacing stuff @kne42 has been thinking about in napari-core.

I also like the "read the docs" standard theme too - 
image
It doesn't have to be what we use for ever, but I think it's a great place to start - lots of people are familiar with it, and I like that navigation bar on the left hand side.

**Comment by sofroniewn** [comment]
All in all though looks like a great start - and will be really helpful to make sure we don't fall behind on our documentation!

File: docs/build/html/_sources/gui.layers.rst.txt
Comment (kne42) [reply docs/build/html/_sources/gui.layers.rst.txt]: They’re accessible in the sense that they’re imported in the corresponding `__init__.py`. Files named with a leading underscore are meant to be "hidden" from users so they logically shouldn’t be present in the docs either. Instead, objects/functions exposed via `__init__.py` should be documented as part of the corresponding module.

Comment (sofroniewn) [reply docs/build/html/_sources/gui.layers.rst.txt]: ah ok nice! thanks for the clarification - that makes sense then. In general I've been struggling a little to know when to `_` functions or not, say in my shapes PR, but we can discuss that over there

Comment (AhmetCanSolak) [reply docs/build/html/_sources/gui.layers.rst.txt]: @kne42 @sofroniewn . These are not meant to be docs for users, IMO. Do you guys want to generate a doc for users? I think we should include all for specs-like docs for developers but should exclude if this will be targeting users.

**Comment by AhmetCanSolak** [comment]
> Have just started looking over the docs, and want to make sure I'm doing it right - I navigated to the docs file and then ran `make html` that then generates a bunch of html files in the `build/html` folder. I then opened up `gui.html` and started clicking around. It looks like this for me
> image

`index.html` is the main page. But you can navigate and find your im sure :)

File: docs/build/html/_modules/gui/_version.html
Comment (AhmetCanSolak) [reply docs/build/html/_modules/gui/_version.html]: @kne42 do you want to see the version somewhere else? or want to get rid of totally?

File: docs/build/html/_sources/gui.layers.rst.txt
Comment (sofroniewn) [reply docs/build/html/_sources/gui.layers.rst.txt]: I'm not really sure about the division between user docs / dev docs. If I go to something like https://pytorch.org/docs/stable/torch.html do I only see user docs? It looks pretty comprehensive.

File: docs/build/html/_modules/gui/_vispy/scene/visuals.html
Comment (AhmetCanSolak) [reply docs/build/html/_modules/gui/_vispy/scene/visuals.html]: Makes sense to hide them but not sure. It is also related with targeted group. users or devs?

File: docs/build/html/_sources/gui.layers.rst.txt
Comment (sofroniewn) [reply docs/build/html/_sources/gui.layers.rst.txt]: seems like things are quite complex right now and there's a lot going on - so i'd maybe lean towards more of the "user" - and not include the `_` files directly or the `vispy` stuff, but then maybe we realize there is certain stuff not present that is important and we need to move over

Comment (AhmetCanSolak) [reply docs/build/html/_sources/gui.layers.rst.txt]: It is targeting users. That example docs doesnt give any implication about design choices/implementation. Showing hidden elements/backports is not required for such usage. It is comprehensive but only giving details for usage of software/API.

Comment (sofroniewn) [reply docs/build/html/_sources/gui.layers.rst.txt]: thanks @AhmetCanSolak - i'd then say we focus on users right now, but open to opinions of others.

**Comment by AhmetCanSolak** [comment]
- docs/build files are removed from source control
- Theme changed to readthedocs
- Doc title is updated

still includes private methods as it is supposed to be developer only docs as we discussed.

**Review by sofroniewn — APPROVED** [review]
I just built the docs successfully and looked at them. Very nice to have the new theme. It's clear though we have a lot of work to do on the actual docs. I'd say we can merge this now, but one other thing that it might be nice to have is an "API Reference section" in the table of contents panel on the left hand side - similar to this screenshot 
image
from the [dask docs](https://dask-ml.readthedocs.io/en/stable/modules/api.html#module-dask_ml.linear_model).
Is that possible @AhmetCanSolak?

**Review by royerloic — DISMISSED** [review]
I would also like to see the "read the docs" standard theme...
After that we should merge.

**Comment by AhmetCanSolak** [comment]
> I would also like to see the "read the docs" standard theme...
> After that we should merge.

right now we have the standard one i guess. Declared on conf.py L81. @royerloic can you let me know if that's not the standard theme?

**Comment by sofroniewn** [comment]
@royerloic we have the standard read the docs theme now - see my screenshot

image

The one thing I'm asking @AhmetCanSolak for before we merge is to have the "python module index" represented as a table of contents with links on the left hand side

**Comment by royerloic** [comment]
Great, let's get this merged asap. 
:-)

**Comment by AhmetCanSolak** [comment]
@sofroniewn i added the reference tree on side bar.

@royerloic once i got approvals i will be merging this

**Comment by royerloic** [comment]
@AhmetCanSolak the CI builds are failing, maybe rebase with the latest master and check...
Otherwise it looks good for merging. Before that, please post here a screenshot for how it looks...

**Comment by sofroniewn** [comment]
I just tried to pull the latest and build the docs but am getting all sorts of errors, as this branch needs to be rebased. Once @AhmetCanSolak has done that I will try again.

Here are some of the errors
```
czipa1osx041:docs nicholassofroniew$ make html
Running Sphinx v1.8.0
loading translations [en]... done
making output directory...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 14 source files that are out of date
updating environment: 14 added, 0 changed, 0 removed
reading sources... [100%] index
WARNING: autodoc: failed to import module '_version' from module 'gui'; the following exception was raised:
No module named 'gui'
WARNING: autodoc: failed to import module 'gui'; the following exception was raised:
No module named 'gui'
```

**Comment by AhmetCanSolak** [comment]
updated my branch. Now docs generated properly. Here SS:
screen shot 2019-03-05 at 11 20 33 am

@sofroniewn those warnings are about docstrings and their formats as far as i see. Let me know if you see anything other than those.

**Comment by sofroniewn** [comment]
This all looks great, thanks for adding. I'm going to dismiss the review from @royerloic and merge!

**Pull request merged**
